src/log/spec.cr

```diff
@@ -0,0 +1,166 @@
+  #
+  # With these methods it is possible to express expected traces in a strict and loosely way, yet always ordered.
+  #
+  # `EntriesChecker#entry` returns the last matched `Entry`. Useful to check additional entry properties other than the message.
+  #
+  # `EntriesChecker#empty` validates there are no pending entries to match.
+  #
+  # Using the yielded `EntriesChecker` allows clearing the entries between statements.
+  #
+  # Invocations can be nested in order to capture each source in their own `EntriesChecker`.
+  #
+  def self.capture(source : String = "*", level : Severity = :info, *, builder = Log.builder)
+    mem_backend = Log::MemoryBackend.new
+    builder.bind(source, level, mem_backend)
+    begin
+      dsl = Log::ArrayEntriesChecker.new(mem_backend.entries).as(EntriesChecker)
+      yield dsl
+      dsl
+    ensure
+      builder.unbind(source, level, mem_backend)
+    end
+  end
+
+  # :ditto:
+  def self.capture(level : Log::Severity = Log::Severity::Info,
+                   *, builder : Log::Builder = Log.builder)
+    capture("*", level, builder: builder) do |dsl|
+      yield dsl
+    end
+  end
+
+  # DSL for `Log.capture`
+  module EntriesChecker
+    # :nodoc:
+    abstract def check(description : String, file = __FILE__, line = __LINE__, & : Entry -> Bool) : self
+
+    # :nodoc:
+    abstract def next(description : String, file = __FILE__, line = __LINE__, & : Entry -> Bool) : self
+
+    # Returns the last entry matched by `#check` or `#next`
+    abstract def entry : Entry
+
+    # Validates that at some point the indicated entry was emitted
+    def check(level : Severity, message : String, file = __FILE__, line = __LINE__) : self
+      self.check("#{level} with #{message.inspect}", file, line) { |e| e.severity == level && e.message == message }
+    end
+
+    # Validates that at some point the indicated entry was emitted
+    def check(level : Severity, pattern : Regex, file = __FILE__, line = __LINE__) : self
+      self.check("#{level} matching #{pattern.inspect}", file, line) { |e| e.severity == level && e.message.matches?(pattern) }
+    end
+
+    # Validates that the indicated entry was the next one to be emitted
+    def next(level : Severity, message : String, file = __FILE__, line = __LINE__) : self
+      self.next("#{level} with #{message.inspect}", file, line) { |e| e.severity == level && e.message == message }
+    end
+
+    # Validates that at some point the indicated entry was emitted
+    def next(level : Severity, pattern : Regex, file = __FILE__, line = __LINE__) : self
+      self.next("#{level} matching #{pattern.inspect}", file, line) { |e| e.severity == level && e.message.matches?(pattern) }
+    end
+
+    # Clears the emitted entries so far
+    abstract def clear
+
+    # Validates that there are no outstanding entries to be validated
+    abstract def empty(file = __FILE__, line = __LINE__)
+  end
+
+  # :nodoc:
+  class ArrayEntriesChecker
```

Comment:
This is `:nodoc:` but also explicitly used in the tests, so it's useful. Should this be called `EntriesChecker` instead and drop the module? Or do you have other implementations in mind?

Reply:
Yeah, I don't see a reason to separate interface and implementation here either.

Reply:
The only need for it was in specs that used `MemoryBackend` explicitly.

We can go that way and make it public. I just wanted to minimize the exposed interface.

The alternative implementations I can think of `ArrayEntriesChecker` is for other testing frameworks. Yet the extensibility is not ready yet.